designs/comprehensions-design.md

```diff
@@ -0,0 +1,174 @@
+
+---
+
+### List Comprehensions
+
+#### Test Case #1
+```python
+print([3 for _ in range(5)])
+```
+Should print ```[3, 3, 3, 3, 3]```
+
+#### Test Case #2
+```python
+print([num * 2 for num in range(5)])
+```
+Should print ```[0, 2, 4, 6, 8]```
+
+#### Test Case #3
+```python
+print([num * 2 if num % 2 == 0 else num * 3 for num in range(5)])
+```
+Should print ```[0, 3, 4, 9, 8]```
+
+#### Test Case #4
+```python
+print([num for num in range(20) if num % 3 == 0])
+```
+Should print ```[0, 3, 6, 9, 12, 15, 18]```
+
+---
+
+### Set/Dictionary Comprehension
+
+#### Test Case #5
+```python
+print({3 for _ in range(5)})
+```
+Should print ```{3}```
+
+#### Test Case #6
+```python
+print({3 if val % 3 == 0 else (2 if val % 2 == 0 else 1) for val in range(8)})
+```
+Should print ```{1, 2, 3}``` (not necessarily sorted)
+
+#### Test Case #7
+```python
+print({val : min(val, 6) for val in range(4, 8)})
+```
+Should print ```{4: 4, 5: 5, 6: 6, 7: 6}```
+
+#### Test Case #8
+```python
+print({val // 2 : val for val in range(8)})
+```
+Should print ```{0: 1, 1: 3, 2: 5, 3: 7}```
+
+---
+
+### Generator
+
+#### Test Case #9
+```python
+gen : generator = None
+gen = (val for val in range(2, 8, 2))
+print(next(gen))
+```
+Should print ```2```
+
+#### Test Case #10
+```python
+gen : generator = None
+x : int = 3
+gen = (val + x for val in range(2, 8) if val < 5)
+print(next(gen))
+print(next(gen))
+print(next(gen))
+print(next(gen))
+```
+Should print
+```python
+5
+6
+7
+StopIteration
+```
+
+---
+---
+
+## Changes to AST, Parser, and Type-checker
+
+We think that comprehension expressions only require changes to the top-level AST -- they can be transformed to the current IR, so here we are only listing changes we intend to make to ```ast.ts```.
+
+---
+
+### AST
+
+```ts
+export type Expr<A> = ...
+    // comprehension expression
+    | { 
+        a?: A, 
+        tag: "comprehension", 
+        type: Type, // type of comprehension - list/set/dict/generator
+        lhs: Expr<A>, 
+        item: string, 
+        iterable: Expr<A>, 
+        ifcond?: Expr<A> 
+    }
+    // ternary expression - in case people want to write this on lhs above
+    | {
+        a?: A, 
+        tag: "ternary", 
+        exprIfTrue: Expr<A>, 
+        ifcond: Expr<A>, 
+        exprIfFalse: Expr<A>
+    }
+```
+
+```ts
+export type Type = ... // assume other groups add list/set/tuple/dict/string here
+    | { 
+        tag: "generator", 
+        type: Type // type of item generated each time calling next()
+    }
```

Comment:
If I want to write a program that takes a generator as an argument to a function, how do I do that?

For example:

```
fun f(g : ???) -> int:
  return next(g)

f((x for x in range(10, 20))
```

This is a good example of a test that doesn't involve lists, strings, etc, and can be something you think about in the first week.